### PR TITLE
docs: enhance documentation support and resolve clippy warnings

### PR DESCRIPTION
**Pull request by JarvisCraft**
# Description

This enhances documentation support by:

* resolving existing docs issues:
  * invalid links;
  * missing `# Safety` sections
* enabling automatic generation of `# Features` section;
* resolving `rustdoc` and `clippy` warnings which were polluting the log.

### CODE REVIEW

**Comment by JarvisCraft** [comment]
It;s worth noting that

```toml
[workspace.package.metadata.docs.rs]
all-features = true
```

in workspace root failed to work (this section seems to be [ignored](https://doc.rust-lang.org/cargo/reference/workspaces.html#the-metadata-table), at least for this specific field).

**Comment by JarvisCraft** [comment]
@str4d, I've merged the latest `main` so that it now should comply with #83.

File: crates/flipperzero/src/furi/string.rs [hunk 1]
```diff
@@ -35,8 +35,8 @@ const WHITESPACE: &[char] = &[
 ///
 /// This is similar to Rust's [`CString`] in that it represents an owned, C-compatible,
 /// nul-terminated string with no nul bytes in the middle. It also has additional methods
-/// to provide the flexibility of Rust's [`String`]. It is used in various APIs of the
-/// Flipper Zero SDK.
+/// to provide the flexibility of Rust's [`String`].
+/// It is used in various APIs of the Flipper Zero SDK.
```
Comment (str4d) [inline crates/flipperzero/src/furi/string.rs:39]: This change is no longer necessary:
```suggestion
/// to provide the flexibility of Rust's [`String`]. It is used in various APIs of the
/// Flipper Zero SDK.
```

File: crates/flipperzero/src/lib.rs [hunk 2]
```diff
@@ -1,10 +1,14 @@
 //! High-level bindings for the Flipper Zero.
-
```
Comment (str4d) [inline crates/flipperzero/src/lib.rs:2]: Please leave this space in (between the documentation and `#![no_std]` line). Note that `//!` is syntactic sugar for `#![doc]`, so the latter should not be treated as part of the module-level block of attributes.

File: crates/flipperzero/src/lib.rs [hunk 3]
```diff
@@ -1,10 +1,14 @@
...
+//!
+//! # Features
+#![cfg_attr(feature = "document-features", doc = document_features::document_features!())]
 #![no_std]
 #![cfg_attr(test, no_main)]
+#![cfg_attr(docsrs, feature(doc_cfg))]
 #![deny(rustdoc::broken_intra_doc_links)]
 
-#[cfg(any(feature = "alloc", docsrs))]
+#[cfg(feature = "alloc")]
```
Comment (str4d) [inline crates/flipperzero/src/lib.rs:10]: This change breaks intra-doc links. Revert it.
```suggestion
#[cfg(any(feature = "alloc", docsrs))]
```

File: crates/sys/src/inlines/furi_hal_gpio.rs [hunk 4]
```diff
@@ -1,13 +1,19 @@
 //! Inlines for Furi HAL GPIO interface.
 //!
-//! See: <https://github.com/flipperdevices/flipperzero-firmware/blob/release/firmware/targets/f7/furi_hal/furi_hal_gpio.h>
+//! See: [furi_hal_gpio.h][1]
```
Comment (str4d) [inline crates/sys/src/inlines/furi_hal_gpio.rs:3]:
```suggestion
//! See: [`furi_hal_gpio.h`][1]
```

File: crates/sys/src/inlines/furi_hal_gpio.rs [hunk 5]
```diff
@@ -1,13 +1,19 @@
...
 use crate as sys;
 
 /// Number of GPIO on one port.
 pub const GPIO_NUMBER: usize = 16;
 
 /// GPIO write pin.
+///
+/// # Safety
+///
+/// Safety invariants are not yet defined for GPIO.
```
Comment (str4d) [inline crates/sys/src/inlines/furi_hal_gpio.rs:16]:
```suggestion
/// `gpio` must be non-null, and the memory it points to must be initialized.
```

File: crates/sys/src/inlines/furi_hal_gpio.rs [hunk 6]
```diff
@@ -17,6 +23,10 @@ pub unsafe extern "C" fn furi_hal_gpio_write(gpio: *const sys::GpioPin, state: b
 }
 
 /// GPIO write pin.
+///
+/// # Safety
+///
+/// Safety invariants are not yet defined for GPIO.
```
Comment (str4d) [inline crates/sys/src/inlines/furi_hal_gpio.rs:29]:
```suggestion
/// `port` must be non-null, and the memory it points to must be initialized.
```

File: crates/sys/src/inlines/furi_hal_gpio.rs [hunk 7]
```diff
@@ -28,6 +38,10 @@ pub unsafe extern "C" fn furi_hal_gpio_write_port_pin(
 }
 
 /// GPIO read pin.
+///
+/// # Safety
+///
+/// Safety invariants are not yet defined for GPIO.
```
Comment (str4d) [inline crates/sys/src/inlines/furi_hal_gpio.rs:44]:
```suggestion
/// `gpio` must be non-null, and the memory it points to must be initialized.
```

File: crates/sys/src/inlines/furi_hal_gpio.rs [hunk 8]
```diff
@@ -37,6 +51,10 @@ pub unsafe extern "C" fn furi_hal_gpio_read(gpio: *const sys::GpioPin) -> bool {
...
```
Comment (str4d) [inline crates/sys/src/inlines/furi_hal_gpio.rs:57]:
```suggestion
/// `port` must be non-null, and the memory it points to must be initialized.
```

File: crates/sys/src/lib.rs [hunk 9]
```diff
@@ -22,8 +22,9 @@ mod inlines;
     non_upper_case_globals,
     non_camel_case_types,
     non_snake_case,
-    clippy::transmute_int_to_bool,
     clippy::useless_transmute,
+    clippy::transmute_int_to_bool,
+    clippy::missing_safety_doc,
```
Comment (str4d) [inline crates/sys/src/lib.rs:27]: Please minimise the diff here:
```suggestion
    clippy::missing_safety_doc,
    clippy::transmute_int_to_bool,
    clippy::useless_transmute,
```

File: tools/src/bin/storage.rs [hunk 10]
```diff
@@ -1,6 +1,8 @@
 //! Storage CLI.
 //!
-//! See: <https://github.com/flipperdevices/flipperzero-firmware/blob/dev/scripts/storage.py>
+//! See: [storage.py][1] script.
```
Comment (str4d) [inline tools/src/bin/storage.rs:3]:
```suggestion
//! See: [`storage.py`][1] script.
```

File: tools/src/bin/storage.rs [hunk 11]
```diff
@@ -1,6 +1,8 @@
...
+//!
+//! [1]: (https://github.com/flipperdevices/flipperzero-firmware/blob/dev/scripts/storage.py)
```
Comment (str4d) [inline tools/src/bin/storage.rs:5]:
```suggestion
//! [1]: https://github.com/flipperdevices/flipperzero-firmware/blob/dev/scripts/storage.py
```

File: crates/flipperzero/Cargo.toml [hunk 12]
```diff
@@ -45,7 +48,17 @@ md-5 = { version = "0.10", default-features = false }
 sha2 = { version = "0.10", default-features = false }
 
 [features]
-# enables features requiring an allocator
+
+#! ## Core features
+
+## Enables features requiring an allocator.
+##
+## This will require you to set up an allocator such as by the following:
```
Comment (str4d) [inline crates/flipperzero/Cargo.toml:56]:
```suggestion
## This will require you to set the global allocator by importing the `flipperzero_alloc`
## crate:
```

File: crates/Cargo.toml [hunk 13]
```diff
@@ -24,6 +24,7 @@ flipperzero-rt = { path = "rt", version = "0.10.0" }
 flipperzero-alloc = { path = "alloc", version = "0.10.0" }
 flipperzero-test = { path = "test", version = "0.10.0" }
 ufmt = "0.2.0"
+document-features = { version = "0.2.0" }
```
Comment (str4d) [inline crates/Cargo.toml:27]:
```suggestion
document-features = "0.2.0"
```

File: crates/flipperzero/src/furi/string.rs [hunk 14]
```diff
@@ -295,63 +295,59 @@ impl FuriString {
...
 
-    /// Returns `true` if the given pattern matches a sub-slice of this string slice.
+    /// Returns `true` if the given pattern `pat` matches a sub-slice of this string slice.
     ///
     /// Returns `false` if it does not.
     ///
-    /// The pattern can be a `&FuriString`, [`c_char`], `&CStr`, [`char`], or a slice of
-    /// [`char`]s.
+    /// The pattern can be a `&FuriString`, [`c_char`], `&CStr`, [`char`],
+    /// or a slice of [`char`]s.
```
Comment (str4d) [inline crates/flipperzero/src/furi/string.rs:303]: Undo this change; the line wrap length was correct for this file (90 characters).
```suggestion
    /// The pattern can be a `&FuriString`, [`c_char`], `&CStr`, [`char`], or a slice of
    /// [`char`]s.
```

File: crates/flipperzero/src/furi/string.rs [hunk 15]
```diff
@@ -295,63 +295,59 @@ impl FuriString {
...
     }
 
-    /// Returns `true` if the given pattern matches a prefix of this string slice.
+    /// Returns `true` if the given pattern `pat` matches a prefix of this string slice.
     ///
     /// Returns `false` if it does not.
     ///
-    /// The pattern can be a `&FuriString`, [`c_char`], `&CStr`, [`char`], or a slice of
-    /// [`char`]s.
+    /// The pattern can be a `&FuriString`, [`c_char`], `&CStr`, [`char`], or a slice of [`char`]s.
```
Comment (str4d) [inline crates/flipperzero/src/furi/string.rs:315]: Ditto:
```suggestion
    /// The pattern can be a `&FuriString`, [`c_char`], `&CStr`, [`char`], or a slice of
    /// [`char`]s.
```

File: crates/flipperzero/src/furi/string.rs [hunk 16]
```diff
@@ -295,63 +295,59 @@ impl FuriString {
...
     }
 
-    /// Returns `true` if the given pattern matches a suffix of this string slice.
+    /// Returns `true` if the given pattern `pat` matches a suffix of this string slice.
     ///
     /// Returns `false` if it does not.
     ///
-    /// The pattern can be a `&FuriString`, [`c_char`], `&CStr`, [`char`], or a slice of
-    /// [`char`]s.
+    /// The pattern can be a `&FuriString`, [`c_char`], `&CStr`, [`char`], or a slice of [`char`]s.
```
Comment (str4d) [inline crates/flipperzero/src/furi/string.rs:326]: Ditto:
```suggestion
    /// The pattern can be a `&FuriString`, [`c_char`], `&CStr`, [`char`], or a slice of
    /// [`char`]s.
```

File: crates/flipperzero/src/furi/string.rs [hunk 17]
```diff
@@ -295,63 +295,59 @@ impl FuriString {
...
     }
 
-    /// Returns the byte index of the first byte of this string that matches the pattern.
+    /// Returns the byte index of the first byte of this string that matches the pattern `pat`.
     ///
     /// Returns [`None`] if the pattern doesn't match.
     ///
-    /// The pattern can be a `&FuriString`, [`c_char`], `&CStr`, [`char`], or a slice of
-    /// [`char`]s.
+    /// The pattern can be a `&FuriString`, [`c_char`], `&CStr`, [`char`], or a slice of [`char`]s.
```
Comment (str4d) [inline crates/flipperzero/src/furi/string.rs:337]: Ditto:
```suggestion
    /// The pattern can be a `&FuriString`, [`c_char`], `&CStr`, [`char`], or a slice of
    /// [`char`]s.
```

File: crates/flipperzero/src/furi/string.rs [hunk 18]
```diff
@@ -295,63 +295,59 @@ impl FuriString {
...
     ///
     /// [`char`]: prim@char
     pub fn ends_with<P: Pattern>(&self, pat: P) -> bool {
         pat.is_suffix_of(self)
     }
 
-    /// Returns the byte index of the first byte of this string that matches the pattern.
+    /// Returns the byte index of the first byte of this string that matches the pattern `pat`.
```
Comment (str4d) [inline crates/flipperzero/src/furi/string.rs:333]:
```suggestion
    /// Returns the byte index of the first byte of this string that matches the pattern
    /// `pat`.
```

File: crates/flipperzero/src/furi/string.rs [hunk 19]
```diff
@@ -295,63 +295,59 @@ impl FuriString {
...
-    /// Returns the byte index for the first byte of the last match of the pattern in this
-    /// string.
+    /// Returns the byte index for the first byte of the last match of the pattern `pat`
+    /// in this string.
     ///
     /// Returns [`None`] if the pattern doesn't match.
     ///
-    /// The pattern can be a `&FuriString`, [`c_char`], `&CStr`, [`char`], or a slice of
-    /// [`char`]s.
+    /// The pattern can be a `&FuriString`, [`c_char`], `&CStr`, [`char`], or a slice of [`char`]s.
```
Comment (str4d) [inline crates/flipperzero/src/furi/string.rs:350]: Ditto:
```suggestion
    /// The pattern can be a `&FuriString`, [`c_char`], `&CStr`, [`char`], or a slice of
    /// [`char`]s.
```

File: crates/flipperzero/src/furi/string.rs [hunk 20]
```diff
@@ -402,19 +398,17 @@ impl FuriString {
 
     /// Repeatedly removes from this string all prefixes and suffixes that match a pattern.
     ///
-    /// The pattern can be a `&FuriString`, [`c_char`], `&CStr`, [`char`], or a slice of
-    /// [`char`]s.
+    /// The pattern can be a `&FuriString`, [`c_char`], `&CStr`, [`char`] or a slice of [`char`]s.
```
Comment (str4d) [inline crates/flipperzero/src/furi/string.rs:401]: Ditto:
```suggestion
    /// The pattern can be a `&FuriString`, [`c_char`], `&CStr`, [`char`], or a slice of
    /// [`char`]s.
```

File: crates/flipperzero/src/furi/string.rs [hunk 21]
```diff
@@ -402,19 +398,17 @@ impl FuriString {
...
         self.trim_start_matches(pat);
         self.trim_end_matches(pat);
     }
 
-    /// Repeatedly removes from this string all prefixes that match a pattern.
+    /// Repeatedly removes from this string all prefixes that match a pattern `pat`.
     ///
-    /// The pattern can be a `&FuriString`, [`c_char`], `&CStr`, [`char`], or a slice of
-    /// [`char`]s.
+    /// The pattern can be a `&FuriString`, [`c_char`], `&CStr`, [`char`] or a slice of [`char`]s.
```
Comment (str4d) [inline crates/flipperzero/src/furi/string.rs:411]: Ditto:
```suggestion
    /// The pattern can be a `&FuriString`, [`c_char`], `&CStr`, [`char`], or a slice of
    /// [`char`]s.
```

File: crates/flipperzero/src/furi/string.rs [hunk 22]
```diff
@@ -428,10 +422,9 @@ impl FuriString {
         while self.strip_prefix(pat) {}
     }
 
-    /// Repeatedly removes from this string all suffixes that match a pattern.
+    /// Repeatedly removes from this string all suffixes that match a pattern `pat`.
     ///
-    /// The pattern can be a `&FuriString`, [`c_char`], `&CStr`, [`char`], or a slice of
-    /// [`char`]s.
+    /// The pattern can be a `&FuriString`, [`c_char`], `&CStr`, [`char`] or a slice of [`char`]s.
```
Comment (str4d) [inline crates/flipperzero/src/furi/string.rs:427]: Ditto:
```suggestion
    /// The pattern can be a `&FuriString`, [`c_char`], `&CStr`, [`char`], or a slice of
    /// [`char`]s.
```

File: crates/flipperzero/src/furi/string.rs [hunk 23]
```diff
@@ -445,31 +438,29 @@ impl FuriString {
...
+    /// Removes the given `prefix` from this string.
     ///
     /// If the string starts with the pattern `prefix`, returns `true`. Unlike
     /// [`Self::trim_start_matches`], this method removes the prefix exactly once.
     ///
     /// If the string does not start with `prefix`, returns `false`.
     ///
-    /// The pattern can be a `&FuriString`, [`c_char`], `&CStr`, [`char`], or a slice of
-    /// [`char`]s.
+    /// The prefix can be a `&FuriString`, [`c_char`], `&CStr`, [`char`], or a slice of [`char`]s.
```
Comment (str4d) [inline crates/flipperzero/src/furi/string.rs:448]:
```suggestion
    /// The prefix can be a `&FuriString`, [`c_char`], `&CStr`, [`char`], or a slice of
    /// [`char`]s.
```

File: crates/flipperzero/src/furi/string.rs [hunk 24]
```diff
@@ -445,31 +438,29 @@ impl FuriString {
...
+    /// Removes the given `suffix` from this string.
     ///
     /// If the string ends with the pattern `suffix`, returns `true`. Unlike
     /// [`Self::trim_end_matches`], this method removes the suffix exactly once.
     ///
     /// If the string does not end with `suffix`, returns `false`.
     ///
-    /// The pattern can be a `&FuriString`, [`c_char`], `&CStr`, [`char`], or a slice of
-    /// [`char`]s.
+    /// The suffix can be a `&FuriString`, [`c_char`], `&CStr`, [`char`], or a slice of [`char`]s.
```
Comment (str4d) [inline crates/flipperzero/src/furi/string.rs:463]:
```suggestion
    /// The suffix can be a `&FuriString`, [`c_char`], `&CStr`, [`char`], or a slice of
    /// [`char`]s.
```

File: crates/flipperzero/src/furi/string/pattern.rs [hunk 25]
```diff
@@ -36,14 +36,17 @@ const FURI_STRING_FAILURE: usize = usize::MAX;
 pub trait Pattern: Sized {
     /// Checks whether the pattern matches anywhere in the haystack.
     #[inline]
+    #[allow(clippy::wrong_self_convention)] // same as in `std`
```
Comment (str4d) [inline crates/flipperzero/src/furi/string/pattern.rs:39]: If these are being moved onto the trait methods, then remove the blanket allow on the trait that I added in #83.

**Review by str4d — CHANGES_REQUESTED** [review]
Generally LGTM

File: crates/flipperzero/src/furi/string.rs (hunk 1 above)
Comment (JarvisCraft) [reply crates/flipperzero/src/furi/string.rs:39]: I think that it is more readable now (when looking at docs in sources) since the new sentence starts from the new line while still maintaining approximately same column width.

File: crates/flipperzero/src/lib.rs (hunk 2 above)
Comment (JarvisCraft) [reply crates/flipperzero/src/lib.rs:2]: I also think that the space fits better, but for some reasong `cargo fmt` prefers to remove the blank line.

File: crates/flipperzero/src/lib.rs (hunk 3 above)
Comment (JarvisCraft) [reply crates/flipperzero/src/lib.rs:10]: I think that it would be better to enable `--all-features` for docs, but I will bring this back as for mow.

File: crates/flipperzero/src/furi/string/pattern.rs (hunk 25 above)
Comment (JarvisCraft) [reply crates/flipperzero/src/furi/string/pattern.rs:39]: Yup, thanks. Didn't see the top-level attribute when resolving the conflicts

File: crates/flipperzero/src/lib.rs (hunk 3 above)
Comment (JarvisCraft) [reply crates/flipperzero/src/lib.rs]: Done

File: crates/flipperzero/src/furi/string.rs (hunk 14 above)
Comment (JarvisCraft) [reply crates/flipperzero/src/furi/string.rs]: Afaik, the [default](https://rust-lang.github.io/rustfmt/?version=v1.5.1&search=#doc_comment_code_block_width) for docomment width is also `100` so we don't even need tro split the lines.

Although if you insist, we could enforce `90` though it does not seems reasonable to me (and, in this speific case, 100 looks better IMO).

Same goes for the other similar comments although you ma

File: crates/flipperzero/src/lib.rs (hunk 2 above)
Comment (str4d) [reply crates/flipperzero/src/lib.rs:2]: Hmm. What if you add a `//!` before the space? That should be valid documentation, and should prevent rustfmt from joining the two.

Comment (JarvisCraft) [reply crates/flipperzero/src/lib.rs:2]: Done

File: crates/flipperzero/src/furi/string.rs (hunk 14 above)
Comment (str4d) [reply crates/flipperzero/src/furi/string.rs]: That default is for code snippets in doc comments, not for documentation itself. I presume it's set to 100 because that is the same break point as for regular code.

I use 90 characters as a break point for documentation because it hits a nice spot for IDE and terminal development readability. 100 for code works well because it's rare you have lines that long, and when the formatter does trigger a break the resulting lines end up much shorter. For documentation, that is not the case; the break point ends up forming a block boundary as there are more granular points to break at, so you end up with mostly full-width lines, and for larger paragraphs 100 is too wide for easy reading. The research done on website UX tends to show that readability drops off at line lengths above 80-85 characters, so including the 4-8 character prefixes on doc lines, 90 sits at a nice break point for readability.

But in any case, what I care about more is minimizing PR diffs. Every change consumes reviewer energy, while they determine if it is semantically meaningful or not. If less energy is spent verifying that a change is just altering formatting, more energy is available for noticing bugs. So I prefer formatting-only changes to be separated out into separate PRs or commits, and ideally automated (though sadly AFAIK `rustfmt` doesn't have even optional support for formatting documentation).

Comment (JarvisCraft) [reply crates/flipperzero/src/furi/string.rs]: A very good point! Thanks for the detailed explanation.

In this case, I will revert the changes to this lines.

Also, let's add this experimental option to the formatting file in the future so that it is more explicit for other committers.

**Comment by JarvisCraft** [comment]
@str4d, I guess there are no more changes here and the CI seems to fail due to some external error

**Comment by str4d** [comment]
Yeah, I've been trying to figure out a way around it, but unfortunately (for good security reasons) GitHub doesn't allow workflows run on PRs from third-party repos to have _any_ write permissions to the base repo. And the Clippy action I found does a small write at the start to create a status entry if it doesn't exist, and _that_ is what is blocking the actual Clippy checks from running. I've opened https://github.com/auguwu/clippy-action/issues/15 upstream for this, but if this can't be addressed easily there, then I'll look into an alternative approach.

File: crates/Cargo.toml [hunk 26]
```diff
@@ -24,6 +24,7 @@ flipperzero-rt = { path = "rt", version = "0.10.0" }
 flipperzero-alloc = { path = "alloc", version = "0.10.0" }
 flipperzero-test = { path = "test", version = "0.10.0" }
 ufmt = "0.2.0"
+document-features ="0.2.0"
```
Comment (str4d) [inline crates/Cargo.toml:27]:
```suggestion
document-features = "0.2.0"
```

File: crates/sys/src/inlines/furi_hal_gpio.rs [hunk 27]
```diff
@@ -37,6 +53,12 @@ pub unsafe extern "C" fn furi_hal_gpio_read(gpio: *const sys::GpioPin) -> bool {
...
+///
```
Comment (str4d) [inline crates/sys/src/inlines/furi_hal_gpio.rs:60]: We've defined the safety invariants we require for this Rust code. Any other requirements relate to the correctness of the inline method's API as called by the SDK.
```suggestion
```

File: crates/sys/src/inlines/furi_hal_gpio.rs [hunk 28]
```diff
@@ -17,6 +23,12 @@ pub unsafe extern "C" fn furi_hal_gpio_write(gpio: *const sys::GpioPin, state: b
 }
 
 /// GPIO write pin.
+///
+/// # Safety
+///
+/// Safety invariants are not yet defined for GPIO.
+///
```
Comment (str4d) [inline crates/sys/src/inlines/furi_hal_gpio.rs:30]:
```suggestion
```

**Review by str4d — APPROVED** [review]
ACK with a few cleanups.
